[compiled autograd] support custom ops backed by c++ autograd::Function

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120681

- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

xmfan committed:
[do not review][compiled autograd] support custom ops backed by c++ autograd::Function

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan commented on test/cpp/jit/test_save_load.cpp:
```diff
@@ -35,7 +35,7 @@ template <class Functor>
   try {
     std::forward<Functor>(functor)();
-  } catch (const Error& e) {
+  } catch (const c10::Error& e) {
```
it's not crystal clear to me why this PR causes the c++ compiler to use torch::autograd::Error here

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -236,6 +245,14 @@ class CompiledNodeArgs {
   }
+  template <typename K, typename V>
+  void collect(ska::flat_hash_map<K, V>& m) {
+    for (auto& [k, v] : m) {
```
1) Need to `collect_size(m.size())`?
2) Is iteration order determisntic?

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -222,6 +224,13 @@ class CompiledNodeArgs {
   template <typename T>
+  void collect(const std::unordered_set<T>& t) {
+    collect_size(t.size());
+    for (const T& i : t) {
```
Is iteration order deterministic?

xmfan replied:
not if the user overrides the default unordered_set. i'll have to sort it

jansel replied:
we will get recompiles for non-determinstic orders

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[do not review][compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Doesn't yet support dynamic shape (will recompile)
- Include files more granularly to avoid namespace pollution and circular imports


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Doesn't yet support dynamic shape (will recompile)
- Include files more granularly to avoid namespace pollution and circular imports


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

current issues:
- test_autograd_cpp_node_saved is flaky, something is probably wrong with the stable sort
- need to collect ivalue properly, maybe using their hash + some handling of list/dict

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jansel commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -173,6 +180,55 @@ struct CppNode : public Node {
+          name());
+    }
+
+    // although neither of the 2 methods below have uniqueness guarantees
+    // it is unlikely for them to collide at the same time
+    args.collect(typeid(T).hash_code());
+    args.collect(std::string(typeid(T).name()));
```
I believe we already specialize on the type of the node here:
https://github.com/pytorch/pytorch/blob/8e4301077efea3265d6707735df5fa56757f6105/torch/csrc/dynamo/compiled_autograd.h#L42-L44

xmfan replied:
I used that initially and it failed the "test_autograd_cpp_node_id" test due to `typeid(CppNode<T1>) == typeid(CppNode<T2>)`. It happens consistently, so I think we need to additionally specialize on the template type

jansel replied:
I see.

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -222,6 +227,17 @@ class CompiledNodeArgs {
+
+    std::vector<T> v(s.begin(), s.end());
+    // stable sort over keys to ensure deterministic collect order
+    std::stable_sort(v.begin(), v.end());
```
Not all types will be sortable.  Which variable is triggering this?  Is it even needed?

xmfan replied:
https://github.com/pytorch/pytorch/blob/368f242e377108a29db9ae821126a9df4b6e40a0/torch/csrc/autograd/custom_function.h#L143-L144

Both are needed because they have public setters, and can be called dynamically in forward
https://github.com/pytorch/pytorch/blob/368f242e377108a29db9ae821126a9df4b6e40a0/torch/csrc/autograd/custom_function.h#L118-L125

jansel replied:
Are these actually used for the needed callbacks?

xmfan replied:
what does "needed callbacks" refer to? not used in fbgemm

jansel replied:
If they aren't used in fbgemm, perhaps just assert they are empty then we won't need this code.

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -236,6 +252,41 @@ class CompiledNodeArgs {
+          return entry.first;
+        });
+    // stable sort to ensure deterministic collect order
+    std::stable_sort(keys.begin(), keys.end());
```
Not all types will be sortable.  Which variable is triggering this?  Is it even needed?

xmfan replied:
https://github.com/pytorch/pytorch/blob/368f242e377108a29db9ae821126a9df4b6e40a0/torch/csrc/autograd/custom_function.h#L111-L113

This one is pretty important, it contains all non-variables, fbgemm example: https://fburl.com/code/1bpdcp0s.

xmfan replied:
This is only implemented for string keys, so it should always be sortable

jansel replied:
Let's add a static_assert (or just change the template) to only support string keys to be safe.

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -236,6 +252,41 @@ class CompiledNodeArgs {
+        collect(entry.value());
+      }
+    } else {
+      collect(at::IValue::hash(iv));
```
What other types hit this case?  Are they all hashable?

xmfan replied:
These are the non-hashable ivalue types:
https://github.com/pytorch/pytorch/blob/368f242e377108a29db9ae821126a9df4b6e40a0/aten/src/ATen/core/ivalue.cpp#L394-L410

From my inspection of internal autograd::Function code, most just need tensors, primitives and list/dict

jansel replied:
Perhaps just turn this into an assert then.  Many of these will be wrong.

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -615,6 +674,36 @@ class SwapSavedVariables {
     }
   }
 
+  template <typename K, typename V>
+  void before(ska::flat_hash_map<K, V>& m) {
+    for (auto& [k, v] : m) {
+      before(k);
+      before(v);
+    }
+  }
+
+  template <typename K, typename V>
+  void after(ska::flat_hash_map<K, V>& m) {
+    for (auto& [k, v] : m) {
+      before(k);
+      before(v);
+    }
+  }
+
+  template <typename T>
+  void before(std::unordered_set<T>& s) {
+    for (auto& i : s) {
+      before(i);
+    }
+  }
+
+  template <typename T>
+  void after(std::unordered_set<T>& s) {
+    for (auto& i : s) {
+      after(i);
+    }
+  }
```
I believe we depend on the order of iteration matching between collect() and before() in some cases.

xmfan replied:
hmm this is not obvious to me, do you have an example? I can add a test for it, and sort or restrict types 

my understanding:
- collect() interacts with the cache key and lifted args
- before() interacts with the stash and lifted args 
- lifted args only has 3 ordered containers: all_size_inputs, dyn_size_inputs and hooks
- but we don't call before() on any of them, so there should be no ordering conflict

jansel replied:
If we create dynamic shapes (because an SymInt changes).

xmfan replied:
i added a sort in case

right now for tensors saved in saved_data (ska::flat_hash_map<std::string, c10::IValue>), we specialize on the memory addresses of tensors impl (which are different even for views).

jansel reviewed (changes requested):
I'm a bit worried about some of the new types being added here.  Though I suspect in the common case we are dealing with:
1) Empty hashmaps
2) Empty sets
3) Etc..

So we might be able to just assert they are empty for safety.

Can you provide examples of the actual values being encountered here?

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -98,6 +99,12 @@ struct TORCH_API Function {
+  // function <T>. Before you can set this flag to true, you need to ensure
+  // that all additional static variables used in the forward and backward
+  // are captured in compiled_args and apply_wih_saved.
+  static constexpr bool is_traceable = false;
```
Please update the comment bloc above the class that we use as documentation to mention this flag.

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -173,6 +180,55 @@ struct CppNode : public Node {
+
+    args.collect(ctx_.saved_data);
+    TORCH_INTERNAL_ASSERT(ctx_.non_differentiable_.empty());
+    TORCH_INTERNAL_ASSERT(ctx_.dirty_inputs_.empty());
```
Should we mention these are not supported in the doc?

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -173,6 +180,55 @@ struct CppNode : public Node {
+    TORCH_INTERNAL_ASSERT(ctx_.non_differentiable_.empty());
+    TORCH_INTERNAL_ASSERT(ctx_.dirty_inputs_.empty());
+    saved.before(ctx_.saved_variables_);
+    saved.before(ctx_.to_save_);
```
This one is missing from the collect?

xmfan replied:
good catch, changed it to assert

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "breaking internal builds, see D54617701" -c ghfirst

```
Stderr:
In file included from fbobjc/Libraries/Odin/OdinSessionTorchC/Exported/TorchModule.mm:7:
In file included from xplat/caffe2/torch/script.h:5:
In file included from xplat/caffe2/torch/csrc/autograd/custom_function.h:10:
xplat/caffe2/torch/csrc/dynamo/compiled_autograd.h:264:24: error: loop variable 'i' binds to a temporary value produced by a range of type 'c10::List<at::IValue>' [-Werror,-Wrange-loop-bind-reference]
      for (const auto& i : list) {
                       ^
xplat/caffe2/torch/csrc/dynamo/compiled_autograd.h:264:12: note: use non-reference type 'ListElementReference<IValue, __wrap_iter<IValue *>>'
      for (const auto& i : list) {
           ^~~~~~~~~~~~~~~
1 error generated.
```

We might have another compiler flag discrepancy between intern and OSS: `-Wrange-loop-bind-reference`.

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

xmfan commented:
@pytorchbot revert -m "breaking internal builds, see D54707287" -c ghfirst

xmfan commented:
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54818488).

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

[ghstack-poisoned]

xmfan commented:
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54818488).

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

[ghstack-poisoned]

xmfan commented:
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54818488).

xmfan committed:
Update on "[compiled autograd] support custom ops backed by c++ autograd::Function"


- Adds support for custom ops backed by c++ custom autograd functions, e.g. fbgemm
- Include files more granularly to avoid namespace pollution and circular imports

limitations:
- requires user to audit their code and opt-in their custom autograd::Function via autograd::Function::is_traceable and maybe additional compiled_args + apply_with_saved implementation. this was the only way I can think of for soundness
- will throw if we can't hash the saved_data i.e. for any non implemented type other than list and dict in at::IValue::hash https://github.com/pytorch/pytorch/blob/b0cfa96e82d7fbd02f5dbcef2632714caf89615d/aten/src/ATen/core/ivalue.cpp#L364
- can technically silently fail if both the typeid hash and the typeid string name of the custom autograd::Function collide at the same time, and an identical autograd graph containing a different custom autograd::Function, yet that has an identical implementation, is called. this case seems extremely unlikely, and the only alternative to hash collision i can think of is compiling with reflection
- tensors not saved via save_variables are not lifted, and are specialized on TensorImpl*'s hash (treated as a memory address). if needed, we can lift them.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54818488](https://our.internmc.facebook.com/intern/diff/D54818488)

[ghstack-poisoned]

xmfan commented:
@xmfan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54818488).

xmfan commented:
@pytorchbot merge

xmfan commented:
@pytorchbot merge -m "manual merge, D54818488 has been landed for 30 mins"

xmfan commented:
manual merge, D54818488 has been landed for 30 mins

@pytorchbot merge
